src/BSON/Int64.c

```diff
@@ -0,0 +1,377 @@
+/*
+ * Copyright 2014-2017 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <php.h>
+#include <Zend/zend_interfaces.h>
+#include <ext/standard/php_var.h>
+#if PHP_VERSION_ID >= 70000
+#include <zend_smart_str.h>
+#else
+#include <ext/standard/php_smart_str.h>
+#endif
+
+#include "phongo_compat.h"
+#include "php_phongo.h"
+#include "php_bson.h"
+
+zend_class_entry* php_phongo_int64_ce;
+
+/* Initialize the object and return whether it was successful. */
+static bool php_phongo_int64_init(php_phongo_int64_t* intern, int64_t integer) /* {{{ */
+{
+	intern->integer     = integer;
+	intern->initialized = true;
+
+	return true;
+} /* }}} */
+
+/* Initialize the object from a numeric string and return whether it was
+ * successful. An exception will be thrown on error. */
+static bool php_phongo_int64_init_from_string(php_phongo_int64_t* intern, const char* s_integer, phongo_zpp_char_len s_integer_len TSRMLS_DC) /* {{{ */
+{
+	int64_t integer;
+	char*   endptr = NULL;
+
+	/* bson_ascii_strtoll() sets errno if conversion fails. If conversion
+	 * succeeds, we still want to ensure that the entire string was parsed. */
+
+	integer = bson_ascii_strtoll(s_integer, &endptr, 10);
+
+	if (errno || (endptr && endptr != ((const char*) s_integer + s_integer_len))) {
+		phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Error parsing \"%s\" as 64-bit integer for %s initialization", s_integer, ZSTR_VAL(php_phongo_int64_ce->name));
+		return false;
+	}
+
+	return php_phongo_int64_init(intern, integer);
+} /* }}} */
+
+/* Initialize the object from a HashTable and return whether it was successful.
+ * An exception will be thrown on error. */
+static bool php_phongo_int64_init_from_hash(php_phongo_int64_t* intern, HashTable* props TSRMLS_DC) /* {{{ */
+{
+#if PHP_VERSION_ID >= 70000
+	zval* value;
+
+	if ((value = zend_hash_str_find(props, "int", sizeof("int") - 1)) && Z_TYPE_P(value) == IS_STRING) {
```

Comment:
It's called ``integer`` in variable name and as property of "intern". To be consistent, the property name in the hash should probably be the same, and not just ``int``. I'm not too fussed about this though.

Reply:
Happy to change. I was trying to be consistent with Decimal128, although I'll admit it has always bugged me that it used "dec" instead of "decimal".